File: 0057-insert-interval/0057-insert-interval.cs
public class Solution {
    public int[][] Insert(int[][] intervals, int[] newInterval) {
        List<int[]> res = new List<int[]>();
        int i = 0;
        
        for (;i < intervals.Length && intervals[i][1] < newInterval[0]; i++)
            res.Add(intervals[i]);
        
        for (;i < intervals.Length && newInterval[0] <= intervals[i][1] && newInterval[1] >= intervals[i][0]; i++)
            newInterval = new int[] { Math.Min(intervals[i][0], newInterval[0]), Math.Max(intervals[i][1], newInterval[1]) };
        
        res.Add(newInterval);
        
        for (;i < intervals.Length; i++)
                res.Add(intervals[i]);
        
        return res.ToArray();
    }
}

File: 0057-insert-interval/0057-insert-interval.scala
object Solution {
  def insert(intervals: Array[Array[Int]], newInterval: Array[Int]): Array[Array[Int]] = {
    val (acc, last) = intervals.foldLeft((Array.empty[Array[Int]], newInterval)) {
      case ((rest, newInterval), head) =>
        if (newInterval.last < head.head) {
          (rest :+ newInterval, head)
        } else if (newInterval.head > head.last) {
          (rest :+ head, newInterval)
        } else {
          (rest, Array(Math.min(head.head, newInterval.head), Math.max(head.last, newInterval.last)))
        }
    }
    acc :+ last
  }
}